src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/runtime/ObjectSynchronizer.java

```diff
@@ -43,8 +43,12 @@ public void update(Observable o, Object data) {
   }
 
   private static synchronized void initialize(TypeDataBase db) throws WrongTypeException {
-    Type type = db.lookupType("ObjectSynchronizer");
-    inUseList = type.getAddressField("_in_use_list").getValue();
+    Type objectSynchronizerType = db.lookupType("ObjectSynchronizer");
+    Type monitorListType = db.lookupType("MonitorList");
+    // Static fields will be left in the default initialized state if the lookup fails
+    Address monitorListAddr = objectSynchronizerType.getField("_in_use_list").getStaticFieldAddress();
+    inUseListHead = monitorListType.getAddressField("_head").getAddress(monitorListAddr);
+    isSupported = true;
```

Comment:
I understand the changes below, but I'm a bit less clear why the above changes were needed. Is this to help simplyfy the iterator logic below and make it easier to check for a null list?

I'm not so sure I like the `isSupported` logic. It seems we should throw an exception at some point if initialize() fails. I think in other classes when some part of initialize() fails, we just allow the NPE to happen when later logic tries to access the value. The check in objectMontitorIterator() was probably a somewhat misguided attempted to be compatible with older VMs, which is not something we really strive for anymore.

Reply:
`inUseListHead` will be null both if `initialize` throws an exception and if the monitor list is null. The previous semantics was to return null if we failed to get the monitor list (because `initialize` failed). But with only  `inUseListHead` we cannot differentiate between the two states. If it is null and `initialize` did not fail, an empty `ObjectMontiorIterator` should be returned instead of null. 

But if it is as you say that this is not a supported use case we could just return an empty `ObjectMontiorIterator` for both cases and rely on the handling of the exception thrown by `initialize`. 

I did not want to change the interface. But I'll follow your initiative here and remove the `null` return from `objectMonitorIterator()`